include/RAJA/exec-agency/forallN_agency.hxx

```diff
@@ -0,0 +1,217 @@
+                                    outer(i, j);
+                                }
+                            }
+                        });
+    }
+};
+
+template <typename Agent, typename  Worker, typename... PREST>
+struct ForallN_Executor<ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        ForallN_PolicyPair<agency_base<Agent, Worker>,
+                                           RangeSegment>,
+                        PREST...> {
+  using PolicyPairType = ForallN_PolicyPair<agency_base<Agent, Worker>, RangeSegment>;
+  PolicyPairType iset_i, iset_j, iset_k;
+
+  using NextExec = ForallN_Executor<PREST...>;
+  NextExec next_exec;
+
+  RAJA_INLINE
+  constexpr ForallN_Executor(
+      PolicyPairType const &iseti_,
+      PolicyPairType const &isetj_,
+      PolicyPairType const &isetk_,
+      PREST const &... prest)
+      : iset_i(iseti_), iset_j(isetj_), iset_k(isetk_), next_exec(prest...)
+  {
+  }
+
+  template <typename BODY>
+  RAJA_INLINE void operator()(BODY body) const
+  {
+    int begin_i = iset_i.getBegin();
+    int begin_j = iset_j.getBegin();
+    int begin_k = iset_k.getBegin();
+    int end_i = iset_i.getEnd();
+    int end_j = iset_j.getEnd();
+    int end_k = iset_k.getEnd();
+
+    ForallN_PeelOuter<NextExec, BODY> outer(next_exec, body);
+
+    auto numThreads = getMaxReduceThreadsCPU();
+    auto tiles = agency::experimental::tile_evenly(
+        agency::experimental::interval(begin_i, end_i), numThreads);
+
+    agency::bulk_invoke(Worker{}(tiles.size()),
+                        [=](Agent& self) {
+                            for (Index_type i : tiles[self.index]) {
+                                for (Index_type j = begin_j; j < end_j; ++j) {
+                                    for (Index_type k = begin_k; k < end_k; ++k) {
+                                        outer(i, j, k);
+                                    }
+                                }
+                            }
+                        });
+    }
+};
+
+/******************************************************************
+ *  forallN_policy()
+ ******************************************************************/
+
+/*!
+ * \brief Tiling policy front-end function.
+ */
+template <typename POLICY, typename BODY, typename Agent, typename Worker, typename... PARGS>
+RAJA_INLINE void forallN_policy(ForallN_Agency_Parallel_Tag<Agent, Worker>,
+                                BODY body,
+                                PARGS... pargs)
+{
+  using NextPolicy = typename POLICY::NextPolicy;
+  using NextPolicyTag = typename POLICY::NextPolicy::PolicyTag;
+
+  // It works if we leave it like this, but not if the part below
+  // is commented out.
+  // This feels weird...
+  forallN_policy<NextPolicy>(NextPolicyTag(), body, pargs...);
```

Comment:
This feels weird, but I think its correct. Basically this function should set things up for the eventual executor, which in our case doesn't take any work. If we did use the part below (commented out) it doesn't work because it ends up spawning an exponential number of threads, which causes problems.